### PR TITLE
Review updates

### PR DESCRIPTION
**Pull request by Eitu33**
Hello,

This PR addresses the review suggestions you sent us, it contains all the changes we made following the review. However, there are some suggestions that we did not apply and we'll explain why here : 

- For the `serialize` method of `Id` (renamed to `to_bytes` now) it’s impossible to return a slice of the object because the byte representation will be created inside the function. We could use a slice as a parameter and have the same behaviour as this crate : https://docs.rs/integer-encoding/latest/integer_encoding/trait.FixedInt.html#tymethod.encode_fixed but we don’t think it’s more suitable for our usage. Open to suggestions.

- For the errors we applied the changes you asked and we now use error types from DB/Felt conversion/Node decoding however we still have errors that hold a String for custom errors and for errors that need a context to be understandable.

- Your idea to use `serde` for the serialization of the changes is great but we think that it's too big of a machinery. The reasons are : 
  -  The conversion has only one purpose and so having the flexibility of the multiple deserialize format given by `serde` etc is useless
  - The conversion is only used in one place for an internal usage
  - This will not optimize the code and can make it more difficult to read

- For the TrieKey (which we have renamed to `KeyOwned`) we can’t use a slice with lifetime because we are saving a collection of these types in the structure of the trie (`MerkleTree`) but we are creating these objects when setting a new key in the trie or removing one. These objects don't live more than the tree because they are created when we manipulate it and so we can’t use a reference to store them in the trie.

- `pub(super)` on the type `Path` prevents us to make the comparison in the tests https://github.com/keep-starknet-strange/bonsai-trie/blob/4ae6caeb59429cee8efd51ebf821569b26caa91e/src/tests/proof.rs#L107

- The `#[test]` attribute already makes the function available only in test mode : https://doc.rust-lang.org/stable/reference/attributes/testing.html#the-test-attribute 

- For the proposition of decoding directly in the key_value db we think it breaks the separation of concern and the simplicity of this structure `KeyValueDB` that is very straightforward for DB interactions. Moreover your `T::decode()` would lead to have a new trait implemented for all the values we want to decode which will lead to complexity and less usability of the `KeyValueDB` structure.

- “I see no reason why you cannot get rid of this and insert directly in self.storage_nodes” https://github.com/keep-starknet-strange/bonsai-trie/blob/87525315abac9611430c5f3c8db922f1d8e0e42f/src/trie/merkle_tree.rs#L464 It’s not possible because we already have a mutual borrow and so we would have two

### CODE REVIEW

File: src/changes.rs [hunk 1]
```diff
@@ -61,20 +31,26 @@ impl ChangeBatch {
         }
     }
 
+    //TODO: Use serde
```
Comment (tdelabro) [inline src/changes.rs]: I think you decided not to use serde? It was in the message coming with your PR.

Comment (tdelabro) [reply src/changes.rs]: Or it is not related to this?

File: src/error.rs [hunk 2]
```diff
@@ -10,5 +19,26 @@ pub enum BonsaiStorageError {
...
     /// Error from the underlying database.
-    Database(String),
+    Database(#[from] DatabaseError),
+    /// Error from Felt conversion
+    Felt252WrapperError(#[from] Felt252WrapperError),
+    /// Error when decoding a node
+    NodeDecodeError(#[from] parity_scale_codec::Error),
+}
+
+impl<DatabaseError> Display for BonsaiStorageError<DatabaseError>
```
Comment (tdelabro) [inline src/error.rs:45]: Thiserror allows you to do that with 
`#[error("Trie error: {0}")]`

Is there a reason you decided not to use it?

File: src/tests/trie_log.rs [hunk 3]
```diff
@@ -183,6 +183,7 @@ fn remove_and_reinsert() {
     bonsai_storage.remove(&bitvec).unwrap();
     bonsai_storage.commit(id1).unwrap();
     let root_hash1 = bonsai_storage.root_hash().unwrap();
+    println!("root_hash1: {:?}", root_hash1);
 
     let id2 = id_builder.new_id();
     println!("before second insert");
```
Comment (tdelabro) [inline src/tests/trie_log.rs]: some forgotten `println`

File: src/trie/path.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+use parity_scale_codec::{Decode, Encode, Error, Input, Output};
+
+use super::{merkle_node::Direction, TrieKey};
+
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Path(pub BitVec<u8, Msb0>);
+
+impl Encode for Path {
+    fn encode_to<T: Output + ?Sized>(&self, dest: &mut T) {
+        // Inspired from scale_bits crate but don't use it to avoid copy and u32 length encoding
```
Comment (tdelabro) [inline src/trie/path.rs]: Can you explain the algo a little bit?
Give an example.

Comment (tdelabro) [reply src/trie/path.rs]: Coz I get the idea that it's compressing the bits. But I don't understand how by reading this code

File: src/trie/path.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+            vec![]
+        } else {
+            [&[path.0.len() as u8], path.0.as_raw_slice()].concat()
+        };
+        TrieKey::Trie(key)
+    }
+}
+
+#[test]
+fn test_shared_path_encode_decode() {
```
Comment (tdelabro) [inline src/trie/path.rs]: Is one single value tested enough?
Using rstest here to go find edge cases could be great

**Review by tdelabro — CHANGES_REQUESTED** [review]

File: src/trie/path.rs (hunk 5 above)
Comment (AurelienFT) [reply src/trie/path.rs]: Done. And it's tested in full tests of the crate everywhere else also.

File: src/trie/path.rs (hunk 4 above)
Comment (AurelienFT) [reply src/trie/path.rs]: I linked the original source code + added an explanation. Tell me you need anything more :)

File: src/tests/trie_log.rs (hunk 3 above)
Comment (AurelienFT) [reply src/tests/trie_log.rs]: Removed.

File: src/error.rs (hunk 2 above)
Comment (AurelienFT) [reply src/error.rs:45]: We didn't know well this crate. No that there is a no_std version and this crate is not compatible we removed his usage and made the conversion directly in the code in a no_std compatible way.

File: src/changes.rs (hunk 1 above)
Comment (AurelienFT) [reply src/changes.rs]: We forget to remove this. Removed

**Review by tdelabro — APPROVED** [review]
